core/lib/src/data/data_stream.rs

```diff
@@ -16,6 +16,43 @@ pub type InnerStream = Chain<Cursor<Vec<u8>>, BodyReader>;
 /// be used as an opaque [`Read`] structure.
 pub struct DataStream(crate InnerStream);
 
+impl DataStream {
+    pub fn read_to_string_with_limit(
+        &mut self,
+        buf: &mut String,
+        limit: usize,
+    ) -> Result<usize, LimitReadError> {
+        self.do_with_limit(limit, |r| r.read_to_string(buf))
+    }
+
+    pub fn read_to_end_with_limit(
+        &mut self,
+        buf: &mut Vec<u8>,
+        limit: usize,
+    ) -> Result<usize, LimitReadError> {
+        self.do_with_limit(limit, |r| r.read_to_end(buf))
+    }
+
+    fn do_with_limit<F: FnOnce(&mut io::Take<&mut Self>) -> io::Result<T>, T>(
```

Comment:
why `Take` and not `&mut dyn Read`?

Reply:
`F: FnOnce(impl Read)` is not currently allowed and a desugaring like `for<R: Read> FnOnce(R)` doesn't exist either as far as I know. Exposing `Take` isn't something we should do if this API were to be made public, so I think `&mut dyn Read` is the only way to erase the type here.